c/parallel/test/test_segmented_reduce.cpp

```diff
@@ -0,0 +1,427 @@
+    cccl_op_t op,
+    cccl_value_t init)
+  {
+    cudaDeviceProp deviceProp;
+    cudaGetDeviceProperties(&deviceProp, 0);
+
+    const int cc_major = deviceProp.major;
+    const int cc_minor = deviceProp.minor;
+
+    const char* cub_path        = TEST_CUB_PATH;
+    const char* thrust_path     = TEST_THRUST_PATH;
+    const char* libcudacxx_path = TEST_LIBCUDACXX_PATH;
+    const char* ctk_path        = TEST_CTK_PATH;
+
+    auto status = cccl_device_segmented_reduce_build(
+      &build,
+      input,
+      output,
+      start_offsets,
+      end_offsets,
+      op,
+      init,
+      cc_major,
+      cc_minor,
+      cub_path,
+      thrust_path,
+      libcudacxx_path,
+      ctk_path);
+
+    REQUIRE(CUDA_SUCCESS == status);
+  }
+
+  void check_sass() const
+  {
+    const std::string sass = inspect_sass(build.cubin, build.cubin_size);
+
+    REQUIRE(sass.find("LDL") == std::string::npos);
+    REQUIRE(sass.find("STL") == std::string::npos);
+  }
+
+  ~SegmentedReduceProgram()
+  {
+    auto status = cccl_device_segmented_reduce_cleanup(&build);
+    REQUIRE(CUDA_SUCCESS == status);
+  }
+
+  CUresult operator()(
+    cccl_iterator_t input,
+    cccl_iterator_t output,
+    unsigned long long num_segments,
+    cccl_iterator_t start_offsets,
+    cccl_iterator_t end_offsets,
+    cccl_op_t op,
+    cccl_value_t init,
+    CUstream stream = 0)
+  {
+    size_t temp_storage_bytes = 0;
+    CUresult status1          = cccl_device_segmented_reduce(
+      build, nullptr, &temp_storage_bytes, input, output, num_segments, start_offsets, end_offsets, op, init, stream);
+
+    if (CUDA_SUCCESS != status1)
+    {
+      return status1;
+    }
+
+    pointer_t<uint8_t> temp_storage(temp_storage_bytes);
+
+    CUresult status2 = cccl_device_segmented_reduce(
+      build,
+      temp_storage.ptr,
+      &temp_storage_bytes,
+      input,
+      output,
+      num_segments,
+      start_offsets,
+      end_offsets,
+      op,
+      init,
+      stream);
+
+    if (CUDA_SUCCESS != status2)
+    {
+      return status2;
+    }
+
+    return status2;
+  }
+};
+
+void segmented_reduce(
+  cccl_iterator_t input,
+  cccl_iterator_t output,
+  unsigned long long num_segments,
+  cccl_iterator_t start_offsets,
+  cccl_iterator_t end_offsets,
+  cccl_op_t op,
+  cccl_value_t init)
+{
+  SegmentedReduceProgram program{input, output, start_offsets, end_offsets, op, init};
+
+  program.check_sass();
+
+  REQUIRE(CUDA_SUCCESS == program(input, output, num_segments, start_offsets, end_offsets, op, init, 0));
+}
+
+using SizeT = unsigned long long;
+
+struct row_offset_iterator_state_t
+{
+  SizeT linear_id;
+  SizeT row_size;
+};
+
+// FIXME: can we cache compiled code for the same TesType and reuse it for different n_rows, n_cols
+using integral_types = std::tuple<std::int32_t, std::int64_t, std::uint32_t, std::uint64_t>;
+TEMPLATE_LIST_TEST_CASE(
+  "segmented_reduce can sum over rows of matrix with integral type", "[segmented_reduce]", integral_types)
```

Comment:
Suggestion: It seems you don't need `integral_types` later, so you can use:
```suggestion
using integral_types = std::tuple<std::int32_t, std::int64_t, std::uint32_t, std::uint64_t>;
TEMPLATE_TEST_CASE(
  "segmented_reduce can sum over rows of matrix with integral type", "[segmented_reduce]", std::int32_t, std::int64_t, std::uint32_t)
```